File: tests/script_verify_rag_2.0.py

import sys
import os
import logging
import time

# Add root
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from config_system import config
from zena_mode.rag_pipeline import LocalRAG

# Configure Logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("RAG2.0_Verify")

def _do_test_rag_upgrade_setup():
    """Helper: setup phase for test_rag_upgrade."""

    logger.info("🚀 Starting RAG 2.0 Verification")

    # 1. Inspect Config
    logger.info(f"Target Model: {config.embedding_config.MODELS['balanced']}")
    logger.info(f"Chunk Strategy: {config.rag.chunk_strategy}")

    # Force settings
    config.rag.chunk_strategy = "semantic"
    config.rag.embedding_model = "balanced" # BGE-Base

    # 2. Initialize RAG (should trigger BGE download/load)
    logger.info("Initializing LocalRAG (this may take time for model download)...")
    rag = LocalRAG()

    # 3. Test Semantic Chunking
    logger.info("Testing Semantic Chunking...")
    sample_text = (
        "Artificial Intelligence is a broad field. Machine learning is a subset of AI. "
        "Deep learning uses neural networks. "
        "The quick brown fox jumps over the lazy dog. "
        "Baking cookies requires flour and sugar. "
        "Preheat the oven to 350 degrees."
    )

    doc = {"content": sample_text, "title": "AI and Cookies", "url": "test://1"}
    chunks = rag.chunk_documents([doc], chunk_size=200, overlap=0)

    for i, c in enumerate(chunks):
        logger.info(f"Chunk {i}: {c['text'][:50]}... ({len(c['text'])} chars)")

    return chunks, rag


def test_rag_upgrade():
    """Test rag upgrade."""
    chunks, rag = _do_test_rag_upgrade_setup()
    # Expect at least 2 chunks (AI vs Cookies) if semantic splitting works well
    if len(chunks) >= 2:
        logger.info("✅ Semantic Chunking looks active (multiple chunks formed)")
    else:
        logger.warning(f"⚠️ Only {len(chunks)} chunks formed. Tune threshold?")

    # 4. Test Semantic Cache
    logger.info("Testing Semantic Cache...")
    rag.cache.clear()
    
    # Ingest chunks manually to index
    rag.add_chunks(chunks)
    
    # Query 1 (Miss)
    start = time.time()
    rag.search("How to bake cookies?", k=1)
    t1 = time.time() - start
    logger.info(f"Query 1 (Miss): {t1:.4f}s")
    
    # Query 2 (Semantic Hit - phrasing diff)
    start = time.time()
    rag.search("Baking instruction", k=1) # Should hit "How to bake cookies?" embedding in cache? 
    # Wait, semantic cache matches QUERY vectors.
    # So "Baking instruction" embedding should be close to "How to bake cookies?" embedding.
    time.time() - start
    
    # Check internals
    if rag.cache._semantic_cache:
        logger.info(f"Cache entries: {len(rag.cache._semantic_cache)}")
        logger.info("✅ Semantic Cache populated")
    else:
        logger.error("❌ Semantic Cache EMPTY")

    logger.info("✅ RAG 2.0 Verification Complete")

if __name__ == "__main__":
    test_rag_upgrade()

File: rust_output/src/script_verify_rag_2.0.rs
use anyhow::{Result, Context};
use crate::config_system::{config};
use crate::rag_pipeline::{LocalRAG};
use std::collections::HashMap;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Helper: setup phase for test_rag_upgrade.
pub fn _do_test_rag_upgrade_setup() -> () {
    // Helper: setup phase for test_rag_upgrade.
    logger.info("🚀 Starting RAG 2.0 Verification".to_string());
    logger.info(format!("Target Model: {}", config::embedding_config.MODELS["balanced".to_string()]));
    logger.info(format!("Chunk Strategy: {}", config::rag.chunk_strategy));
    config::rag.chunk_strategy = "semantic".to_string();
    config::rag.embedding_model = "balanced".to_string();
    logger.info("Initializing LocalRAG (this may take time for model download)...".to_string());
    let mut rag = LocalRAG();
    logger.info("Testing Semantic Chunking...".to_string());
    let mut sample_text = "Artificial Intelligence is a broad field. Machine learning is a subset of AI. Deep learning uses neural networks. The quick brown fox jumps over the lazy dog. Baking cookies requires flour and sugar. Preheat the oven to 350 degrees.".to_string();
    let mut doc = HashMap::from([("content".to_string(), sample_text), ("title".to_string(), "AI and Cookies".to_string()), ("url".to_string(), "test://1".to_string())]);
    let mut chunks = rag.chunk_documents(vec![doc], /* chunk_size= */ 200, /* overlap= */ 0);
    for (i, c) in chunks.iter().enumerate().iter() {
        logger.info(format!("Chunk {}: {}... ({} chars)", i, c["text".to_string()][..50], c["text".to_string()].len()));
    }
    (chunks, rag)
}

/// Test rag upgrade.
pub fn test_rag_upgrade() -> () {
    // Test rag upgrade.
    let (mut chunks, mut rag) = _do_test_rag_upgrade_setup();
    if chunks.len() >= 2 {
        logger.info("✅ Semantic Chunking looks active (multiple chunks formed)".to_string());
    } else {
        logger.warning(format!("⚠️ Only {} chunks formed. Tune threshold?", chunks.len()));
    }
    logger.info("Testing Semantic Cache...".to_string());
    rag.cache::clear();
    rag.add_chunks(chunks);
    let mut start = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
    rag.search("How to bake cookies?".to_string(), /* k= */ 1);
    let mut t1 = (std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - start);
    logger.info(format!("Query 1 (Miss): {:.4}s", t1));
    let mut start = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
    rag.search("Baking instruction".to_string(), /* k= */ 1);
    (std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - start);
    if rag.cache::_semantic_cache {
        logger.info(format!("Cache entries: {}", rag.cache::_semantic_cache.len()));
        logger.info("✅ Semantic Cache populated".to_string());
    } else {
        logger.error("❌ Semantic Cache EMPTY".to_string());
    }
    logger.info("✅ RAG 2.0 Verification Complete".to_string());
}
